src/hotspot/share/prims/jvmtiRawMonitor.cpp

```diff
@@ -410,38 +373,41 @@ int JvmtiRawMonitor::raw_wait(jlong millis, Thread* self) {
   OrderAccess::fence();
 
   intptr_t save = _recursions;
-  _recursions = 0;
-  _waiters++;
-  ret = simple_wait(self, millis);
-  _recursions = save;
-  _waiters--;
-
-  guarantee(self == _owner, "invariant");
 
   if (self->is_Java_thread()) {
     JavaThread* jt = self->as_Java_thread();
     guarantee(jt->thread_state() == _thread_in_native, "invariant");
-    for (;;) {
-      if (!SafepointMechanism::should_process(jt)) {
-        // Not suspended so we're done here:
-        break;
+
+    _recursions = 0;
+    ret = simple_wait(self, millis);
```

Comment:
IMHO the guarantee at L379 is redundant with the newly added identical guarantee in `JvmtiRawMonitor::simple_wait()` at L240.

In case you agree to remove the guarantee, I don't see why the following statements cannot be pulled out of the if-statement.
```
    _recursions = 0;
    ret = simple_wait(self, millis);
    _recursions = save;
```

Reply:
Yes, it can be re-written as something below, but is this more readable?

```
  _recursions = 0;                                                                                               
  ret = simple_wait(self, millis);                                                                                   
  
  if (self->is_Java_thread()) { // JavaThread re-enter                                                                                        
    JavaThread* jt = self->as_Java_thread();                                                                           
    {                                                                                                                  
      ThreadInVMfromNative tivmfn(jt);                                                                                 
      for (;;) {                                                                                                       
        ExitOnSuspend eos(this);                                                                                       
        {                                                                                                              
          ThreadBlockInVMPreprocess<ExitOnSuspend> tbivmp(jt, eos);                                                    
          simple_enter(jt);                                                                                            
        }                                                                                                              
        if (!eos.monitor_exited()) {                                                                                   
          break;                                                                                                       
        }                                                                                                              
      }                                                                                                                
      if (jt->is_interrupted(true)) {                                                                                  
        ret = M_INTERRUPTED;                                                                                           
      }                                                                                                                
    }                                                                                                                  
  } else { // Non-JavaThread re-enter                                                                 
    assert(ret != M_INTERRUPTED, "Only JavaThreads can be interrupted");                                               
    simple_enter(self);                                                                                            
  }
  _recursions = save;                                                                                              

```

Reply:
Fixed